metadata-ingestion-modules/airflow-plugin/src/datahub_airflow_plugin/datahub_listener.py

```diff
@@ -0,0 +1,411 @@
+        sql_parsing_result = None
+        if task_metadata:
+            sql_parsing_result: Optional[
+                SqlParsingResult
+            ] = task_metadata.run_facets.pop("datahub_sql", None)
+        if sql_parsing_result:
+            if sql_parsing_result.debug_info.error:
+                datajob.properties["datahub_sql_parser_error"] = str(
+                    sql_parsing_result.debug_info.error
+                )
+            else:
+                input_urns.extend(sql_parsing_result.in_tables)
+                output_urns.extend(sql_parsing_result.out_tables)
+
+                if sql_parsing_result.column_lineage:
+                    fine_grained_lineages.extend(
+                        FineGrainedLineageClass(
+                            upstreamType=FineGrainedLineageUpstreamTypeClass.FIELD_SET,
+                            downstreamType=FineGrainedLineageDownstreamTypeClass.FIELD,
+                            upstreams=[
+                                builder.make_schema_field_urn(
+                                    upstream.table, upstream.column
+                                )
+                                for upstream in column_lineage.upstreams
+                            ],
+                            downstreams=[
+                                builder.make_schema_field_urn(
+                                    downstream.table, downstream.column
+                                )
+                                for downstream in [column_lineage.downstream]
+                                if downstream.table
+                            ],
+                        )
+                        for column_lineage in sql_parsing_result.column_lineage
+                    )
+
+        # Add DataHub-native inlets/outlets.
+        # These are filtered out by the extractor, so we need to add them manually.
+        input_urns.extend(
+            iolet.urn for iolet in get_task_inlets(task) if isinstance(iolet, _Entity)
+        )
+        output_urns.extend(
+            iolet.urn for iolet in get_task_outlets(task) if isinstance(iolet, _Entity)
+        )
+
+        # Write the lineage to the datajob object.
+        datajob.inlets.extend(input_urns)
+        datajob.outlets.extend(output_urns)
+        datajob.fine_grained_lineages.extend(fine_grained_lineages)
+
+        # Merge in extra stuff that was present in the DataJob we constructed
+        # at the start of the task.
+        if complete:
+            original_datajob = self._datajob_holder.get(str(datajob.urn), None)
+        else:
+            self._datajob_holder[str(datajob.urn)] = datajob
+            original_datajob = None
+
+        if original_datajob:
+            logger.debug("Merging start datajob into finish datajob")
+            datajob.inlets.extend(original_datajob.inlets)
+            datajob.outlets.extend(original_datajob.outlets)
+            datajob.fine_grained_lineages.extend(original_datajob.fine_grained_lineages)
```

Comment:
How come you build up local vars `input_urns`, `output_urns`, and `fine_grained_lineages` above, but then directly update the datajob here? Could we be consistent here?